spark/src/main/scala/geotrellis/spark/io/CollectionLayerReader.scala

```diff
@@ -0,0 +1,80 @@
+package geotrellis.spark.io
+
+import geotrellis.spark._
+import geotrellis.spark.io.avro._
+import geotrellis.spark.io.json._
+import geotrellis.util._
+import spray.json._
+
+import scalaz.std.vector._
+import scalaz.concurrent.{Strategy, Task}
+import scalaz.stream.{Process, nondeterminism}
+import java.util.concurrent.{ExecutorService, Executors}
+
+import scala.reflect._
+
+abstract class CollectionLayerReader[ID] { self =>
+  def read[
+    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag,
+    V: AvroRecordCodec: ClassTag,
+    M: JsonFormat: GetComponent[?, Bounds[K]]
+  ](id: ID, rasterQuery: LayerQuery[K, M], indexFilterOnly: Boolean): Seq[(K, V)] with Metadata[M]
+
+  def reader[
+    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag,
+    V: AvroRecordCodec: ClassTag,
+    M: JsonFormat: GetComponent[?, Bounds[K]]
+  ]: Reader[ID, Seq[(K, V)] with Metadata[M]] =
+    new Reader[ID, Seq[(K, V)] with Metadata[M]] {
+      def read(id: ID): Seq[(K, V)] with Metadata[M] =
+        self.read[K, V, M](id)
+    }
+
+  def read[
+    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag,
+    V: AvroRecordCodec: ClassTag,
+    M: JsonFormat: GetComponent[?, Bounds[K]]
+  ](id: ID, rasterQuery: LayerQuery[K, M]): Seq[(K, V)] with Metadata[M] =
+    read[K, V, M](id, rasterQuery, false)
+
+  def read[
+    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag,
+    V: AvroRecordCodec: ClassTag,
+    M: JsonFormat: GetComponent[?, Bounds[K]]
+  ](id: ID): Seq[(K, V)] with Metadata[M] =
+    read[K, V, M](id, new LayerQuery[K, M])
+
+  def query[
+    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag,
+    V: AvroRecordCodec: ClassTag,
+    M: JsonFormat: GetComponent[?, Bounds[K]]
+  ](layerId: ID): BoundLayerQuery[K, M, Seq[(K, V)] with Metadata[M]] =
+    new BoundLayerQuery(new LayerQuery, read[K, V, M](layerId, _))
+}
+
+object CollectionLayerReader {
+  def njoin[K: AvroRecordCodec: Boundable, V: AvroRecordCodec](
```

Comment:
We as for `AvroRecordCodec` but we do not use them. If the `readFunc` is `Long => Array[Byte]` we can spin the iterator here and do the decoding. You'll need an extra `refinementFilter: K => Boolean` to capture the `filterIndexOnly` param.